In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('view.csv')

In [3]:
df

Source Location ID     City        State     Country  Latitude  \
0     NSRDB       20013  Jodhpur    Rajasthan       India     26.35   
1       NaN         NaN      NaN          NaN         NaN       NaN   
2      Hour         GHI      DHI  Temperature  Wind Speed       NaN   
3         1           0        0           11         0.2       NaN   
4         2           0        0           10         0.2       NaN   
...     ...         ...      ...          ...         ...       ...   
8758   8756           0        0           15         0.3       NaN   
8759   8757           0        0           14         0.3       NaN   
8760   8758           0        0           13         0.3       NaN   
8761   8759           0        0           12         0.3       NaN   
8762   8760           0        0           11         0.2       NaN   

      Longitude  Time Zone  Elevation  Local Time Zone       DHI Units  \
0         73.05        5.5        0.0              5.5            w/m2   
1           NaN        NaN        NaN              NaN             NaN   
2           NaN        NaN        NaN              NaN             NaN   
3           NaN        NaN        NaN              NaN             NaN   
4           NaN        NaN        NaN              NaN  Abbreviations:   
...         ...        ...        ...              ...             ...   
8758        NaN        NaN        NaN              NaN             NaN   
8759        NaN        NaN        NaN              NaN             NaN   
8760        NaN        NaN        NaN              NaN             NaN   
8761        NaN        NaN        NaN              NaN             NaN   
8762        NaN        NaN        NaN              NaN             NaN   

     DNI Units GHI Units Temperature Units Wind Speed Units  
0         w/m2      w/m2                 c              m/s  
1          NaN       NaN               NaN              NaN  
2          NaN       NaN               NaN              NaN  
3          NaN       NaN               NaN              NaN  
4          NaN       NaN               NaN              NaN  
...        ...       ...               ...              ...  
8758       NaN       NaN               NaN              NaN  
8759       NaN       NaN               NaN              NaN  
8760       NaN       NaN               NaN              NaN  
8761       NaN       NaN               NaN              NaN  
8762       NaN       NaN               NaN              NaN  

[8763 rows x 15 columns]

## Data Cleaning

In [4]:
df.drop(df.loc[:, 'Latitude':'Wind Speed Units'].columns, axis=1, inplace=True)
#Dropping unnecessary columns

In [5]:
df.dtypes
#Checking if the columns are iun the ideal type or not

Source         object
Location ID    object
City           object
State          object
Country        object
dtype: object

In [6]:
df.rename(columns={'Source':'Hour', 'Location ID':'GHI', 'City':'DHI', 'State':'Temperature', 'Country':'Wind Speed'}, inplace=True)
#Renaming the columns to our need

In [7]:
df.drop(df.index[:3], inplace=True)
df = df.reset_index(drop=True)
#Dropping first three unnecessary rows and resetting the index

In [8]:
for i in df.columns:
    df[i]=df[i].astype('float32')
#Changing the data types to float to ease calculations in the future

In [9]:
df

Hour  GHI  DHI  Temperature  Wind Speed
0        1.0  0.0  0.0         11.0         0.2
1        2.0  0.0  0.0         10.0         0.2
2        3.0  0.0  0.0         10.0         0.2
3        4.0  0.0  0.0          9.0         0.2
4        5.0  0.0  0.0          9.0         0.2
...      ...  ...  ...          ...         ...
8755  8756.0  0.0  0.0         15.0         0.3
8756  8757.0  0.0  0.0         14.0         0.3
8757  8758.0  0.0  0.0         13.0         0.3
8758  8759.0  0.0  0.0         12.0         0.3
8759  8760.0  0.0  0.0         11.0         0.2

[8760 rows x 5 columns]

## Finding BHI

In [10]:
df['BHI']=df['GHI']-df['DHI']
#Finding beam radiation I_b from I_g and I_d

In [11]:
Ib=df['BHI'].to_numpy()
Id=df['DHI'].to_numpy()
Ig=df['GHI'].to_numpy()
#Creating arrays as numpy arrays are computationally faster than pandas

## Finding I_sc'

In [12]:
day=np.arange(1,366)
df['day']=np.repeat(day,24)
#Creating the days of an year and saving them to the dataframe

In [13]:
I_sc_day = 1367*(1+0.033*np.cos(np.deg2rad(360*day/365)))
#Finding the daywise solar irradiance

In [14]:
I_sc_day=np.repeat(I_sc_day, 24)
df['I_sc\'']=I_sc_day
#As solar irradiance is constant for a day, we repeat the value for 24 hours and save it to the dataframe

In [15]:
pd.set_option('display.max_columns', None)

In [16]:
df[172*24:173*24]

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'
4128  4129.0    0.0    0.0         32.0         0.6    0.0  173  1322.490874
4129  4130.0    0.0    0.0         31.0         0.6    0.0  173  1322.490874
4130  4131.0    0.0    0.0         30.0         0.6    0.0  173  1322.490874
4131  4132.0    0.0    0.0         29.0         0.6    0.0  173  1322.490874
4132  4133.0    0.0    0.0         29.0         0.6    0.0  173  1322.490874
4133  4134.0    0.0    0.0         29.0         0.5    0.0  173  1322.490874
4134  4135.0   54.0   51.0         30.0         0.5    3.0  173  1322.490874
4135  4136.0  245.0  152.0         32.0         0.5   93.0  173  1322.490874
4136  4137.0  453.0  192.0         34.0         0.5  261.0  173  1322.490874
4137  4138.0  658.0  238.0         37.0         0.5  420.0  173  1322.490874
4138  4139.0  816.0  262.0         39.0         0.5  554.0  173  1322.490874
4139  4140.0  922.0  276.0         41.0         0.5  646.0  173  1322.490874
4140  4141.0  965.0  281.0         43.0         0.5  684.0  173  1322.490874
4141  4142.0  943.0  278.0         43.0         0.5  665.0  173  1322.490874
4142  4143.0  858.0  267.0         43.0         0.5  591.0  173  1322.490874
4143  4144.0  703.0  235.0         43.0         0.5  468.0  173  1322.490874
4144  4145.0  530.0  215.0         41.0         0.5  315.0  173  1322.490874
4145  4146.0  313.0  173.0         40.0         0.5  140.0  173  1322.490874
4146  4147.0  102.0   87.0         38.0         0.5   15.0  173  1322.490874
4147  4148.0    0.0    0.0         37.0         0.5    0.0  173  1322.490874
4148  4149.0    0.0    0.0         36.0         0.5    0.0  173  1322.490874
4149  4150.0    0.0    0.0         34.0         0.5    0.0  173  1322.490874
4150  4151.0    0.0    0.0         33.0         0.6    0.0  173  1322.490874
4151  4152.0    0.0    0.0         32.0         0.6    0.0  173  1322.490874

## Finding Solar Zenith Angle and incidence angle

θ = f(β, γ, δ, ω, φ)<br>
θz = f(β=0, γ, δ, ω, φ)<br>
where β=collector tilt angle, γ=surface azimuth angle, δ=declination angle, ω=hour angle, φ=latitude<br>
Latitude is given as 26.35<br>Assuming that the panels are facing south<br>Assuming an average value of tilt angle (for now)

In [17]:
φ=26+35/60
γ=0
β=15

### Declination Angle

In [18]:
δ=23.45*np.sin(np.deg2rad(360*(284+day)/365))
#Finding the declination angle using the formula

In [19]:
δ=np.repeat(δ, 24)
df['δ']=δ
#As solar declination angle is constant for a day, we repeat the value for 24 hours and save it to the dataframe

In [20]:
hr = np.arange(0,24*365)%24+1
df['daywise_hours'] = hr
#We would want daywise hours from 1-24 to be clear about our daywise radiation

### Hour Angle

In [21]:
E=[]
for i in day:
    B=(i-1)*360/365
    E.append(229.18*(0.000075+0.001868*np.cos(np.deg2rad(B))-0.032077*np.sin(np.deg2rad(B))-0.014615*np.cos(np.deg2rad(2*B))-0.04089*np.sin(np.deg2rad(2*B))))
E=np.repeat(E,24)
df['Error in min']=E
#Finding daywise error in time and saving it to the dataframe

In [22]:
LAT=[]
for i,j in enumerate(hr):
    LAT.append(j*60-37.8+E[i])
LAT=np.asarray(LAT)
df['LAT_minutes']=LAT
#Finding LAT in minutes and saving it to the dataframe

In [23]:
df['LAT']=pd.to_datetime(df.LAT_minutes, unit='m').dt.strftime('%H:%M')
#Finding LAT in time format and saving it to the dataframe

In [24]:
ω=(LAT-720)/60*15
df['ω']=ω
#Finding hour angle in time format and saving it to the dataframe

In [25]:
df

Hour  GHI  DHI  Temperature  Wind Speed  BHI  day        I_sc'  \
0        1.0  0.0  0.0         11.0         0.2  0.0    1  1412.104316   
1        2.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
2        3.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
3        4.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
4        5.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
...      ...  ...  ...          ...         ...  ...  ...          ...   
8755  8756.0  0.0  0.0         15.0         0.3  0.0  365  1412.111000   
8756  8757.0  0.0  0.0         14.0         0.3  0.0  365  1412.111000   
8757  8758.0  0.0  0.0         13.0         0.3  0.0  365  1412.111000   
8758  8759.0  0.0  0.0         12.0         0.3  0.0  365  1412.111000   
8759  8760.0  0.0  0.0         11.0         0.2  0.0  365  1412.111000   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  
0    -23.011637              1     -2.904169    19.295831  00:19 -175.176042  
1    -23.011637              2     -2.904169    79.295831  01:19 -160.176042  
2    -23.011637              3     -2.904169   139.295831  02:19 -145.176042  
3    -23.011637              4     -2.904169   199.295831  03:19 -130.176042  
4    -23.011637              5     -2.904169   259.295831  04:19 -115.176042  
...         ...            ...           ...          ...    ...         ...  
8755 -23.085911             20     -2.453134  1159.746866  19:19  109.936716  
8756 -23.085911             21     -2.453134  1219.746866  20:19  124.936716  
8757 -23.085911             22     -2.453134  1279.746866  21:19  139.936716  
8758 -23.085911             23     -2.453134  1339.746866  22:19  154.936716  
8759 -23.085911             24     -2.453134  1399.746866  23:19  169.936716  

[8760 rows x 14 columns]

In [26]:
def value(β):
    first = np.sin(np.deg2rad(β))*np.cos(np.deg2rad(γ))*(np.cos(np.deg2rad(δ))*np.cos(np.deg2rad(ω))*np.sin(np.deg2rad(φ))-np.sin(np.deg2rad(δ))*np.cos(np.deg2rad(φ)))
    second = np.sin(np.deg2rad(β))*np.sin(np.deg2rad(γ))*np.cos(np.deg2rad(δ))*np.sin(np.deg2rad(ω))
    third = np.cos(np.deg2rad(β))*(np.cos(np.deg2rad(δ))*np.cos(np.deg2rad(ω))*np.cos(np.deg2rad(φ))+np.sin(np.deg2rad(δ))*np.sin(np.deg2rad(φ)))                                                                                                    
    return first+second+third
#Formula for cosθ and cosθz with β varying

In [27]:
cosθ=value(15)
df['cosθ']=cosθ
#Incidence angles and saving to the dataframe

In [28]:
cosθz=value(0)
df['cosθz']=cosθz
#Solar zenith angles and saving to the dataframe

### Radiation on Tilted Surface - Isotropic Model (liu Jordan)

In [29]:
rb=cosθ/cosθz
df['rb']=rb
#Finding the tilt factor for beam radiation and saving to the dataframe

In [30]:
max(rb)
#Weird value noted meansd more data cleaning needed, will do after finding It

45025.013979347415

In [31]:
rd=(1+np.cos(np.deg2rad(β)))/2
#Finding the tilt factor for diffused radiation

In [32]:
rr=0.2*(1-np.cos(np.deg2rad(β)))/2
#Finding the tilt factor for reflected radiation

In [33]:
It_LJ = np.multiply(Ib,rb) + Id*rd + Ig*rr
df['It_LJ']=It_LJ
#Finding the tilted radiation and saving to the dataframe

### Radiation on Tilted Surface - Anisotropic Model (HDKR model)

In [34]:
Io=I_sc_day*cosθz

In [35]:
Ai=df['BHI']/Io

In [36]:
It_HDRK=(Ib+Id*Ai)*rb+Id*(1-Ai)*rd*(1+np.sqrt(np.divide(Ib, Ig, out=np.zeros_like(Ib), where=Ig!=0))*(np.sin(np.deg2rad(β/2)))**3)+Ig*rr

In [37]:
df['It_HDRK']=It_HDRK

In [38]:
df[:50]

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
0    1.0    0.0    0.0         11.0         0.2    0.0    1  1412.104316   
1    2.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
2    3.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
3    4.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
4    5.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
5    6.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
6    7.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
7    8.0    0.0    0.0         12.0         0.3    0.0    1  1412.104316   
8    9.0  158.0   79.0         15.0         0.2   79.0    1  1412.104316   
9   10.0  359.0  120.0         19.0         0.2  239.0    1  1412.104316   
10  11.0  529.0  144.0         24.0         0.2  385.0    1  1412.104316   
11  12.0  644.0  157.0         28.0         0.2  487.0    1  1412.104316   
12  13.0  693.0  162.0         30.0         0.1  531.0    1  1412.104316   
13  14.0  671.0  159.0         30.0         0.1  512.0    1  1412.104316   
14  15.0  579.0  149.0         30.0         0.1  430.0    1  1412.104316   
15  16.0  427.0  130.0         28.0         0.1  297.0    1  1412.104316   
16  17.0  234.0   96.0         25.0         0.1  138.0    1  1412.104316   
17  18.0   26.0   26.0         22.0         0.2    0.0    1  1412.104316   
18  19.0    0.0    0.0         21.0         0.2    0.0    1  1412.104316   
19  20.0    0.0    0.0         20.0         0.3    0.0    1  1412.104316   
20  21.0    0.0    0.0         19.0         0.3    0.0    1  1412.104316   
21  22.0    0.0    0.0         18.0         0.3    0.0    1  1412.104316   
22  23.0    0.0    0.0         17.0         0.3    0.0    1  1412.104316   
23  24.0    0.0    0.0         16.0         0.3    0.0    1  1412.104316   
24  25.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
25  26.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
26  27.0    0.0    0.0         14.0         0.3    0.0    2  1412.084267   
27  28.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
28  29.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
29  30.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
30  31.0    0.0    0.0         12.0         0.3    0.0    2  1412.084267   
31  32.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
32  33.0  157.0   79.0         16.0         0.3   78.0    2  1412.084267   
33  34.0  333.0  129.0         20.0         0.2  204.0    2  1412.084267   
34  35.0  509.0  145.0         25.0         0.2  364.0    2  1412.084267   
35  36.0  611.0  169.0         29.0         0.1  442.0    2  1412.084267   
36  37.0  682.0  163.0         30.0         0.1  519.0    2  1412.084267   
37  38.0  667.0  159.0         31.0         0.0  508.0    2  1412.084267   
38  39.0  576.0  149.0         30.0         0.0  427.0    2  1412.084267   
39  40.0  425.0  130.0         28.0         0.0  295.0    2  1412.084267   
40  41.0  234.0   96.0         25.0         0.1  138.0    2  1412.084267   
41  42.0   29.0   28.0         22.0         0.2    1.0    2  1412.084267   
42  43.0    0.0    0.0         21.0         0.2    0.0    2  1412.084267   
43  44.0    0.0    0.0         20.0         0.3    0.0    2  1412.084267   
44  45.0    0.0    0.0         19.0         0.3    0.0    2  1412.084267   
45  46.0    0.0    0.0         18.0         0.3    0.0    2  1412.084267   
46  47.0    0.0    0.0         17.0         0.3    0.0    2  1412.084267   
47  48.0    0.0    0.0         16.0         0.3    0.0    2  1412.084267   
48  49.0    0.0    0.0         16.0         0.3    0.0    3  1412.050859   
49  50.0    0.0    0.0         15.0         0.2    0.0    3  1412.050859   

            δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
0  -23.011637              1     -2.904169    19.

The difference between It_LJ and It_HDRK can be seen to be less than 10%

In [39]:
df['Ibn']=df['BHI']/df['cosθz']
#Finding the beam normal radiation

In [40]:
df

Hour  GHI  DHI  Temperature  Wind Speed  BHI  day        I_sc'  \
0        1.0  0.0  0.0         11.0         0.2  0.0    1  1412.104316   
1        2.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
2        3.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
3        4.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
4        5.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
...      ...  ...  ...          ...         ...  ...  ...          ...   
8755  8756.0  0.0  0.0         15.0         0.3  0.0  365  1412.111000   
8756  8757.0  0.0  0.0         14.0         0.3  0.0  365  1412.111000   
8757  8758.0  0.0  0.0         13.0         0.3  0.0  365  1412.111000   
8758  8759.0  0.0  0.0         12.0         0.3  0.0  365  1412.111000   
8759  8760.0  0.0  0.0         11.0         0.2  0.0  365  1412.111000   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
0    -23.011637              1     -2.904169    19.295831  00:19 -175.176042   
1    -23.011637              2     -2.904169    79.295831  01:19 -160.176042   
2    -23.011637              3     -2.904169   139.295831  02:19 -145.176042   
3    -23.011637              4     -2.904169   199.295831  03:19 -130.176042   
4    -23.011637              5     -2.904169   259.295831  04:19 -115.176042   
...         ...            ...           ...          ...    ...         ...   
8755 -23.085911             20     -2.453134  1159.746866  19:19  109.936716   
8756 -23.085911             21     -2.453134  1219.746866  20:19  124.936716   
8757 -23.085911             22     -2.453134  1279.746866  21:19  139.936716   
8758 -23.085911             23     -2.453134  1339.746866  22:19  154.936716   
8759 -23.085911             24     -2.453134  1399.746866  23:19  169.936716   

          cosθ     cosθz        rb  It_LJ  It_HDRK  Ibn  
0    -0.976979 -0.995142  0.981749    0.0      0.0 -0.0  
1    -0.926739 -0.949279  0.976256    0.0      0.0 -0.0  
2    -0.818692 -0.850645  0.962437    0.0      0.0 -0.0  
3    -0.660202 -0.705963  0.935179    0.0      0.0 -0.0  
4    -0.462069 -0.525092  0.879977    0.0      0.0 -0.0  
...        ...       ...       ...    ...      ...  ...  
8755 -0.386020 -0.455984  0.846565    0.0      0.0 -0.0  
8756 -0.594815 -0.646588  0.919929    0.0      0.0 -0.0  
8757 -0.768439 -0.805085  0.954481    0.0      0.0 -0.0  
8758 -0.895061 -0.920675  0.972178    0.0      0.0 -0.0  
8759 -0.966051 -0.985481  0.980284    0.0      0.0 -0.0  

[8760 rows x 20 columns]

## More Data Cleaning

In [41]:
df.sort_values(by='It_LJ', ascending=False)

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
305    306.0   51.0   44.0         17.0         0.3    7.0   13  1410.986136   
2004  2005.0  949.0  250.0         37.0         0.2  699.0   84  1372.615384   
1908  1909.0  937.0  249.0         39.0         0.1  688.0   80  1375.681683   
2700  2701.0  983.0  294.0         44.0         0.1  689.0  113  1350.501891   
2724  2725.0  984.0  294.0         45.0         0.0  690.0  114  1349.781618   
...      ...    ...    ...          ...         ...    ...  ...          ...   
3787  3788.0    0.0    0.0         37.0         0.5    0.0  158  1325.841862   
3773  3774.0    0.0    0.0         31.0         0.4    0.0  158  1325.841862   
233    234.0   53.0   43.0         22.0         0.1   10.0   10  1411.444264   
257    258.0   50.0   43.0         23.0         0.2    7.0   11  1411.304668   
281    282.0   54.0   47.0         21.0         0.3    7.0   12  1411.151944   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
305  -21.596777             18     -7.901680  1034.298320  17:14   78.574580   
2004   1.210481             13     -6.610441   735.589559  12:15    3.897390   
1908  -0.403653             13     -7.861940   734.338060  12:14    3.584515   
2700  12.274096             13      1.650186   743.850186  12:23    5.962547   
2724  12.616221             13      1.851597   744.051597  12:24    6.012899   
...         ...            ...           ...          ...    ...         ...   
3787  22.747999             20      1.576165  1163.776165  19:23  110.944041   
3773  22.747999              6      1.576165   323.776165  05:23  -99.055959   
233  -22.039625             18     -6.741642  1035.458358  17:15   78.864589   
257  -21.898483             18     -7.136171  1035.063829  17:15   78.765957   
281  -21.750852             18     -7.522965  1034.677035  17:14   78.669259   

          cosθ     cosθz            rb          It_LJ       It_HDRK  \
305   0.106528  0.000002  45025.013979  315218.522000  4.154272e+09   
2004  0.981392  0.901471      1.088656    1009.945167  1.025076e+03   
1908  0.976278  0.889360      1.097731    1003.189568  1.019464e+03   
2700  0.994749  0.964248      1.031631    1003.134413  1.010958e+03   
2724  0.994578  0.965633      1.029975    1003.026740  1.010597e+03   
...        ...       ...           ...            ...           ...   
3787 -0.245295 -0.121763      2.014520       0.000000  0.000000e+00   
3773 -0.064557  0.043228     -1.493411       0.000000  0.000000e+00   
233   0.100022 -0.007832    -12.770624     -85.258238  4.497967e+02   
257   0.102191 -0.005249    -19.466760     -93.829540  7.370764e+02   
281   0.104360 -0.002637    -39.568309    -230.594906  3.354120e+03   

               Ibn  
305   2.958610e+06  
2004  7.753995e+02  
1908  7.735899e+02  
2700  7.145462e+02  
2724  7.145571e+02  
...            ...  
3787 -0.000000e+00  
3773  0.000000e+00  
233  -1.276775e+03  
257  -1.333463e+03  
281  -2.654075e+03  

[8760 rows x 20 columns]

Clearly rb max value is 45025 and It max value is coming out to be 315218 W/m2 which is way out of bound, so we replace it with zero. Negative values of It are also replaced with zero

In [42]:
It_LJ[It_LJ<0]=0
It_LJ[It_LJ>1100]=0
df['It_LJ']=It_LJ

In [43]:
It_HDRK[It_HDRK<0]=0
It_HDRK[It_HDRK>1100]=0
df['It_HDRK']=It_HDRK

I_t is cleaned now<br>
Let's clean I_bn

In [44]:
Ibn=np.asarray(df['Ibn'])
Ibn[Ibn<0]=0
Ibn[Ibn>1100]=0
df['Ibn']=Ibn

In [45]:
Ibn[6833]

895.9043249935204

In [46]:
df.sort_values(by='Ibn', ascending=False)

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
638    639.0  653.0  141.0         24.0         0.3  512.0   27  1407.325562   
6833  6834.0   85.0   61.0         29.0         0.1   24.0  285  1375.681683   
711    712.0  519.0  129.0         29.0         0.2  390.0   30  1406.228047   
663    664.0  512.0  128.0         23.0         0.3  384.0   28  1406.971532   
709    710.0  768.0  169.0         30.0         0.2  599.0   30  1406.228047   
...      ...    ...    ...          ...         ...    ...  ...          ...   
3719  3720.0    0.0    0.0         35.0         0.6    0.0  155  1326.849966   
3718  3719.0    0.0    0.0         36.0         0.5    0.0  155  1326.849966   
3717  3718.0    0.0    0.0         38.0         0.5    0.0  155  1326.849966   
3716  3717.0    0.0    0.0         39.0         0.4    0.0  155  1326.849966   
8759  8760.0    0.0    0.0         11.0         0.2    0.0  365  1412.111000   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
638  -18.791918             15    -12.185457   850.014543  14:10   32.503636   
6833  -8.482187             18     13.711457  1055.911457  17:35   83.977864   
711  -18.042778             16    -12.818059   909.381941  15:09   47.345485   
663  -18.547676             16    -12.408511   909.791489  15:09   47.447872   
709  -18.042778             14    -12.818059   789.381941  13:09   17.345485   
...         ...            ...           ...          ...    ...         ...   
3719  22.423707             24      2.102571  1404.302571  23:24  171.075643   
3718  22.423707             23      2.102571  1344.302571  22:24  156.075643   
3717  22.423707             22      2.102571  1284.302571  21:24  141.075643   
3716  22.423707             21      2.102571  1224.302571  20:24  126.075643   
8759 -23.085911             24     -2.453134  1399.746866  23:19  169.936716   

          cosθ     cosθz        rb       It_LJ     It_HDRK         Ibn  
638   0.717460  0.569845  1.259044  785.453435  810.404899  898.490518  
6833  0.072034  0.026789  2.688999  124.786354  192.584859  895.904325  
711   0.568944  0.437546  1.300307  635.690532  661.728136  891.334555  
663   0.564207  0.431011  1.309032  630.232072  656.749731  890.929149  
709   0.826911  0.673038  1.228625  904.684053  931.079672  889.994702  
...        ...       ...       ...         ...         ...         ...  
3719 -0.818006 -0.645959  1.266344    0.000000    0.000000   -0.000000  
3718 -0.751164 -0.584941  1.284172    0.000000    0.000000   -0.000000  
3717 -0.627912 -0.472427  1.329122    0.000000    0.000000   -0.000000  
3716 -0.456650 -0.316085  1.444706    0.000000    0.000000   -0.000000  
8759 -0.966051 -0.985481  0.980284    0.000000    0.000000   -0.000000  

[8760 rows x 20 columns]

In [47]:
df
#Final dataframe

Hour  GHI  DHI  Temperature  Wind Speed  BHI  day        I_sc'  \
0        1.0  0.0  0.0         11.0         0.2  0.0    1  1412.104316   
1        2.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
2        3.0  0.0  0.0         10.0         0.2  0.0    1  1412.104316   
3        4.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
4        5.0  0.0  0.0          9.0         0.2  0.0    1  1412.104316   
...      ...  ...  ...          ...         ...  ...  ...          ...   
8755  8756.0  0.0  0.0         15.0         0.3  0.0  365  1412.111000   
8756  8757.0  0.0  0.0         14.0         0.3  0.0  365  1412.111000   
8757  8758.0  0.0  0.0         13.0         0.3  0.0  365  1412.111000   
8758  8759.0  0.0  0.0         12.0         0.3  0.0  365  1412.111000   
8759  8760.0  0.0  0.0         11.0         0.2  0.0  365  1412.111000   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
0    -23.011637              1     -2.904169    19.295831  00:19 -175.176042   
1    -23.011637              2     -2.904169    79.295831  01:19 -160.176042   
2    -23.011637              3     -2.904169   139.295831  02:19 -145.176042   
3    -23.011637              4     -2.904169   199.295831  03:19 -130.176042   
4    -23.011637              5     -2.904169   259.295831  04:19 -115.176042   
...         ...            ...           ...          ...    ...         ...   
8755 -23.085911             20     -2.453134  1159.746866  19:19  109.936716   
8756 -23.085911             21     -2.453134  1219.746866  20:19  124.936716   
8757 -23.085911             22     -2.453134  1279.746866  21:19  139.936716   
8758 -23.085911             23     -2.453134  1339.746866  22:19  154.936716   
8759 -23.085911             24     -2.453134  1399.746866  23:19  169.936716   

          cosθ     cosθz        rb  It_LJ  It_HDRK  Ibn  
0    -0.976979 -0.995142  0.981749    0.0      0.0 -0.0  
1    -0.926739 -0.949279  0.976256    0.0      0.0 -0.0  
2    -0.818692 -0.850645  0.962437    0.0      0.0 -0.0  
3    -0.660202 -0.705963  0.935179    0.0      0.0 -0.0  
4    -0.462069 -0.525092  0.879977    0.0      0.0 -0.0  
...        ...       ...       ...    ...      ...  ...  
8755 -0.386020 -0.455984  0.846565    0.0      0.0 -0.0  
8756 -0.594815 -0.646588  0.919929    0.0      0.0 -0.0  
8757 -0.768439 -0.805085  0.954481    0.0      0.0 -0.0  
8758 -0.895061 -0.920675  0.972178    0.0      0.0 -0.0  
8759 -0.966051 -0.985481  0.980284    0.0      0.0 -0.0  

[8760 rows x 20 columns]

In [48]:
#df.to_excel('data.xlsx')

# Modelling

In [49]:
from CoolProp.CoolProp import PhaseSI, PropsSI, get_global_param_string
import CoolProp.CoolProp as cp

In [50]:
h7= cp.PropsSI('H', 'T',90+273,'P', 150000,'Water')
h8= cp.PropsSI('H', 'T',10+273,'P', 100000,'Water')

In [51]:
T10=5+273
T11=85+273

Assuming ideal heat exchanger <br>
Using m2Cpw(T7-T8)=m3Cpm(T11-T10) we find m2

In [52]:
from sympy import symbols, solve

In [53]:
m = symbols('x')

In [54]:
m3=4.166*1032

In [55]:
expr=m*(h7-h8)-m3*3.93*(T11-T10)*1000

In [56]:
mww=solve(expr)[0]/1000
print("The value of the volume flow rate of water in the heat exchanger in m3/hr is", mww)

The value of the volume flow rate of water in the heat exchanger in m3/hr is 4.03515210571784


In [57]:
print("Useful energy needed for pasteurization is", m3*3.93*(T11-T10), "in kJ/hr")
print("Useful energy needed for pasteurization is", m3*3.93*(T11-T10)/(3600*1000), "in MW")

Useful energy needed for pasteurization is 1351703.6928000003 in kJ/hr
Useful energy needed for pasteurization is 0.3754732480000001 in MW


We'll model for June 21st

In [58]:
n=173

In [59]:
dff=df.iloc[(n-1)*24:n*24]
#Selecting the values for June 21st

In [60]:
dff

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
4128  4129.0    0.0    0.0         32.0         0.6    0.0  173  1322.490874   
4129  4130.0    0.0    0.0         31.0         0.6    0.0  173  1322.490874   
4130  4131.0    0.0    0.0         30.0         0.6    0.0  173  1322.490874   
4131  4132.0    0.0    0.0         29.0         0.6    0.0  173  1322.490874   
4132  4133.0    0.0    0.0         29.0         0.6    0.0  173  1322.490874   
4133  4134.0    0.0    0.0         29.0         0.5    0.0  173  1322.490874   
4134  4135.0   54.0   51.0         30.0         0.5    3.0  173  1322.490874   
4135  4136.0  245.0  152.0         32.0         0.5   93.0  173  1322.490874   
4136  4137.0  453.0  192.0         34.0         0.5  261.0  173  1322.490874   
4137  4138.0  658.0  238.0         37.0         0.5  420.0  173  1322.490874   
4138  4139.0  816.0  262.0         39.0         0.5  554.0  173  1322.490874   
4139  4140.0  922.0  276.0         41.0         0.5  646.0  173  1322.490874   
4140  4141.0  965.0  281.0         43.0         0.5  684.0  173  1322.490874   
4141  4142.0  943.0  278.0         43.0         0.5  665.0  173  1322.490874   
4142  4143.0  858.0  267.0         43.0         0.5  591.0  173  1322.490874   
4143  4144.0  703.0  235.0         43.0         0.5  468.0  173  1322.490874   
4144  4145.0  530.0  215.0         41.0         0.5  315.0  173  1322.490874   
4145  4146.0  313.0  173.0         40.0         0.5  140.0  173  1322.490874   
4146  4147.0  102.0   87.0         38.0         0.5   15.0  173  1322.490874   
4147  4148.0    0.0    0.0         37.0         0.5    0.0  173  1322.490874   
4148  4149.0    0.0    0.0         36.0         0.5    0.0  173  1322.490874   
4149  4150.0    0.0    0.0         34.0         0.5    0.0  173  1322.490874   
4150  4151.0    0.0    0.0         33.0         0.6    0.0  173  1322.490874   
4151  4152.0    0.0    0.0         32.0         0.6    0.0  173  1322.490874   

              δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
4128  23.448046              1     -1.544107    20.655893  00:20 -174.836027   
4129  23.448046              2     -1.544107    80.655893  01:20 -159.836027   
4130  23.448046              3     -1.544107   140.655893  02:20 -144.836027   
4131  23.448046              4     -1.544107   200.655893  03:20 -129.836027   
4132  23.448046              5     -1.544107   260.655893  04:20 -114.836027   
4133  23.448046              6     -1.544107   320.655893  05:20  -99.836027   
4134  23.448046              7     -1.544107   380.655893  06:20  -84.836027   
4135  23.448046              8     -1.544107   440.655893  07:20  -69.836027   
4136  23.448046              9     -1.544107   500.655893  08:20  -54.836027   
4137  23.448046             10     -1.544107   560.655893  09:20  -39.836027   
4138  23.448046             11     -1.544107   620.655893  10:20  -24.836027   
4139  23.448046             12     -1.544107   680.655893  11:20   -9.836027   
4140  23.448046             13     -1.544107   740.655893  12:20    5.163973   
4141  23.448046             14     -1.544107   800.655893  13:20   20.163973   
4142  23.448046             15     -1.544107   860.655893  14:20   35.163973   
4143  23.448046             16     -1.544107   920.655893  15:20   50.163973   
4144  23.448046             17     -1.544107   980.655893  16:20   65.163973   
4145  23.448046             18     -1.544107  1040.655893  17:20   80.163973   
4146  23.448046             19     -1.544107  1100.655893  18:20   95.163973   
4147  23.448046             20     -1.544107  1160.655893  19:20  110.163973   
4148  23.448046             21     -1.544107  1220.655893  20:20  125.163973   
4149  23.448046             22     -1.544107  1280.655893  21:20  140.163973   
4150  23.448046             23     -1.544107  1340.655893  22:20  155.163973   
4151  23.448046             24     -1.544107  1400.655893  23:20  170.163973   

      

At LAT close to 12 noon, 12:20PM, Ibn is maximum which is 687.3W/m2

In [61]:
I_bn=687.3
#Qu_dot = F'A(S-U_L(T_fm-T_a)) where S = Ibn*Ta_eff
#Qu_dot = m3*Cp_milk*(T11-T10)
#Equating both the equations in the above two comments gives us area needed
A=(m3*3.93*(T11-T10)*1000)/(0.89*(I_bn*0.8-7.5*25)*3600)
print(A)
# Returns area needed for the solar field for that time instant

1164.3209525103061


In [62]:
mw1=mww*Ibn/I_bn
df['mw1']=mw1
#Finding the mass flow rates for all the other hours based on these values

In [63]:
SM=1.2
print("The area of solar field needed is", SM*A)
#Final area requirement for the whole solar field

The area of solar field needed is 1397.1851430123672


Close the solar plant when Ibn<200Wm-2<br>
Finding the corresponding mass flow rate and closing it accordingly<br>
So, if volume flow rate is lower than a specofoc value Ill close the solar plant.

In [64]:
m_cric = mww*200/I_bn
(df['mw1']<m_cric).sum()

5536

Clearly, we need a reheater for almost the 5536 hours in a year

In [65]:
mw_fromsolar=np.where(mw1<m_cric, 0, mw1).tolist()
df['mw_fromsolar']=mw_fromsolar

In [66]:
df[5000:5050]

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
5000  5001.0  404.0  217.0         29.0         0.4  187.0  209  1326.501898   
5001  5002.0  403.0  324.0         31.0         0.4   79.0  209  1326.501898   
5002  5003.0  619.0  362.0         32.0         0.4  257.0  209  1326.501898   
5003  5004.0  737.0  373.0         34.0         0.4  364.0  209  1326.501898   
5004  5005.0  858.0  292.0         35.0         0.4  566.0  209  1326.501898   
5005  5006.0  785.0  357.0         35.0         0.4  428.0  209  1326.501898   
5006  5007.0  654.0  390.0         35.0         0.4  264.0  209  1326.501898   
5007  5008.0  447.0  353.0         34.0         0.4   94.0  209  1326.501898   
5008  5009.0  387.0  282.0         33.0         0.4  105.0  209  1326.501898   
5009  5010.0  251.0  196.0         32.0         0.4   55.0  209  1326.501898   
5010  5011.0   87.0   80.0         31.0         0.4    7.0  209  1326.501898   
5011  5012.0    0.0    0.0         30.0         0.4    0.0  209  1326.501898   
5012  5013.0    0.0    0.0         29.0         0.4    0.0  209  1326.501898   
5013  5014.0    0.0    0.0         29.0         0.4    0.0  209  1326.501898   
5014  5015.0    0.0    0.0         28.0         0.4    0.0  209  1326.501898   
5015  5016.0    0.0    0.0         28.0         0.5    0.0  209  1326.501898   
5016  5017.0    0.0    0.0         27.0         0.5    0.0  210  1326.849966   
5017  5018.0    0.0    0.0         27.0         0.4    0.0  210  1326.849966   
5018  5019.0    0.0    0.0         27.0         0.4    0.0  210  1326.849966   
5019  5020.0    0.0    0.0         26.0         0.4    0.0  210  1326.849966   
5020  5021.0    0.0    0.0         26.0         0.4    0.0  210  1326.849966   
5021  5022.0    0.0    0.0         26.0         0.4    0.0  210  1326.849966   
5022  5023.0   20.0   20.0         27.0         0.4    0.0  210  1326.849966   
5023  5024.0  168.0  130.0         28.0         0.4   38.0  210  1326.849966   
5024  5025.0  360.0  223.0         30.0         0.4  137.0  210  1326.849966   
5025  5026.0  345.0  311.0         31.0         0.4   34.0  210  1326.849966   
5026  5027.0  424.0  383.0         33.0         0.4   41.0  210  1326.849966   
5027  5028.0  697.0  372.0         33.0         0.4  325.0  210  1326.849966   
5028  5029.0  622.0  446.0         34.0         0.4  176.0  210  1326.849966   
5029  5030.0  622.0  480.0         33.0         0.4  142.0  210  1326.849966   
5030  5031.0  532.0  415.0         33.0         0.4  117.0  210  1326.849966   
5031  5032.0  336.0  314.0         32.0         0.4   22.0  210  1326.849966   
5032  5033.0  311.0  264.0         31.0         0.4   47.0  210  1326.849966   
5033  5034.0  142.0  140.0         30.0         0.3    2.0  210  1326.849966   
5034  5035.0   59.0   59.0         29.0         0.3    0.0  210  1326.849966   
5035  5036.0    0.0    0.0         29.0         0.3    0.0  210  1326.849966   
5036  5037.0    0.0    0.0         29.0         0.4    0.0  210  1326.849966   
5037  5038.0    0.0    0.0         28.0         0.4    0.0  210  1326.849966   
5038  5039.0    0.0    0.0         28.0         0.4    0.0  210  1326.849966   
5039  5040.0    0.0    0.0         28.0         0.5    0.0  210  1326.849966   
5040  5041.0    0.0    0.0         27.0         0.5    0.0  211  1327.209932   
5041  5042.0    0.0    0.0         27.0         0.5    0.0  211  1327.209932   
5042  5043.0    0.0    0.0         27.0         0.5    0.0  211  1327.209932   
5043  5044.0    0.0    0.0         27.0         0.4    0.0  211  1327.209932   
5044  5045.0    0.0    0.0         26.0         0.4    0.0  211  1327.209932   
5045  5046.0    0.0    0.0         26.0         0.4    0.0  211  1327.209932   
5046  5047.0   22.0   22.0         27.0         0.4    0.0  211  1327.209932   
5047  5048.0  119.0  114.0         28.0         0.4    5.0  211  1327.209932   
5048  5049.0  301.0  237.0         29.0         0.4   64.0  211  1327.209932   
5049  50

The above line shows how the values of mass flow rate in the dataframe are replaced with zeros when the flow rate is low, or in other words, when the DNI is low.<br>
Now, let's check if there are any times when the flow rate from the solar field is higher than required in the HX and we'll regulate it by passing it to the storage tank and using it during a later time

In [67]:
(df['mw1']>mww).sum()

1049

In [68]:
mHX=np.repeat(mww,8760)

In [69]:
mdif=mHX-mw_fromsolar

In [70]:
df['m_extras']=np.where(mdif>0,0,mdif).tolist()

In [71]:
df['m_fromheater']=np.where(mdif<0,0,mdif).tolist()

In [72]:
df[:50]

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
0    1.0    0.0    0.0         11.0         0.2    0.0    1  1412.104316   
1    2.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
2    3.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
3    4.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
4    5.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
5    6.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
6    7.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
7    8.0    0.0    0.0         12.0         0.3    0.0    1  1412.104316   
8    9.0  158.0   79.0         15.0         0.2   79.0    1  1412.104316   
9   10.0  359.0  120.0         19.0         0.2  239.0    1  1412.104316   
10  11.0  529.0  144.0         24.0         0.2  385.0    1  1412.104316   
11  12.0  644.0  157.0         28.0         0.2  487.0    1  1412.104316   
12  13.0  693.0  162.0         30.0         0.1  531.0    1  1412.104316   
13  14.0  671.0  159.0         30.0         0.1  512.0    1  1412.104316   
14  15.0  579.0  149.0         30.0         0.1  430.0    1  1412.104316   
15  16.0  427.0  130.0         28.0         0.1  297.0    1  1412.104316   
16  17.0  234.0   96.0         25.0         0.1  138.0    1  1412.104316   
17  18.0   26.0   26.0         22.0         0.2    0.0    1  1412.104316   
18  19.0    0.0    0.0         21.0         0.2    0.0    1  1412.104316   
19  20.0    0.0    0.0         20.0         0.3    0.0    1  1412.104316   
20  21.0    0.0    0.0         19.0         0.3    0.0    1  1412.104316   
21  22.0    0.0    0.0         18.0         0.3    0.0    1  1412.104316   
22  23.0    0.0    0.0         17.0         0.3    0.0    1  1412.104316   
23  24.0    0.0    0.0         16.0         0.3    0.0    1  1412.104316   
24  25.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
25  26.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
26  27.0    0.0    0.0         14.0         0.3    0.0    2  1412.084267   
27  28.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
28  29.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
29  30.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
30  31.0    0.0    0.0         12.0         0.3    0.0    2  1412.084267   
31  32.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
32  33.0  157.0   79.0         16.0         0.3   78.0    2  1412.084267   
33  34.0  333.0  129.0         20.0         0.2  204.0    2  1412.084267   
34  35.0  509.0  145.0         25.0         0.2  364.0    2  1412.084267   
35  36.0  611.0  169.0         29.0         0.1  442.0    2  1412.084267   
36  37.0  682.0  163.0         30.0         0.1  519.0    2  1412.084267   
37  38.0  667.0  159.0         31.0         0.0  508.0    2  1412.084267   
38  39.0  576.0  149.0         30.0         0.0  427.0    2  1412.084267   
39  40.0  425.0  130.0         28.0         0.0  295.0    2  1412.084267   
40  41.0  234.0   96.0         25.0         0.1  138.0    2  1412.084267   
41  42.0   29.0   28.0         22.0         0.2    1.0    2  1412.084267   
42  43.0    0.0    0.0         21.0         0.2    0.0    2  1412.084267   
43  44.0    0.0    0.0         20.0         0.3    0.0    2  1412.084267   
44  45.0    0.0    0.0         19.0         0.3    0.0    2  1412.084267   
45  46.0    0.0    0.0         18.0         0.3    0.0    2  1412.084267   
46  47.0    0.0    0.0         17.0         0.3    0.0    2  1412.084267   
47  48.0    0.0    0.0         16.0         0.3    0.0    2  1412.084267   
48  49.0    0.0    0.0         16.0         0.3    0.0    3  1412.050859   
49  50.0    0.0    0.0         15.0         0.2    0.0    3  1412.050859   

            δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
0  -23.011637              1     -2.904169    19.

The fact that the m_extras are negative depict the storage low rates in that specifc hour of the day

In [73]:
amy=[]

In [74]:
for i in range(1,366):
    dff=[]
    dff=df.iloc[(i-1)*24:i*24]
    amy.append(np.absolute(dff['m_extras']).sum())

Array amy gives the extra mass flow rate we get every day of the year

In [75]:
alice=[]

for i in range(1,366):
    dfn=[]
    dfn=df.iloc[(i-1)*24+18]
    if dfn['m_fromheater']>amy[i-1]:
        alice.append('yes')
    else:
        alice.append('no')

np.unique(alice, return_counts=True)

(array(['no', 'yes'], dtype='<U3'), array([ 29, 336]))

So, on 29 days, at 7:00PM local time, we have more storage from that day than needed at 7:00PM
And on 336 days, at 7:00PM local time, we'd need more water form heater than that stored during solar field operational times.

In [76]:
df['EnergyfromSolar(inJ)']=df['mw_fromsolar']*4184*80*1000
df[:50]

Hour    GHI    DHI  Temperature  Wind Speed    BHI  day        I_sc'  \
0    1.0    0.0    0.0         11.0         0.2    0.0    1  1412.104316   
1    2.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
2    3.0    0.0    0.0         10.0         0.2    0.0    1  1412.104316   
3    4.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
4    5.0    0.0    0.0          9.0         0.2    0.0    1  1412.104316   
5    6.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
6    7.0    0.0    0.0         11.0         0.3    0.0    1  1412.104316   
7    8.0    0.0    0.0         12.0         0.3    0.0    1  1412.104316   
8    9.0  158.0   79.0         15.0         0.2   79.0    1  1412.104316   
9   10.0  359.0  120.0         19.0         0.2  239.0    1  1412.104316   
10  11.0  529.0  144.0         24.0         0.2  385.0    1  1412.104316   
11  12.0  644.0  157.0         28.0         0.2  487.0    1  1412.104316   
12  13.0  693.0  162.0         30.0         0.1  531.0    1  1412.104316   
13  14.0  671.0  159.0         30.0         0.1  512.0    1  1412.104316   
14  15.0  579.0  149.0         30.0         0.1  430.0    1  1412.104316   
15  16.0  427.0  130.0         28.0         0.1  297.0    1  1412.104316   
16  17.0  234.0   96.0         25.0         0.1  138.0    1  1412.104316   
17  18.0   26.0   26.0         22.0         0.2    0.0    1  1412.104316   
18  19.0    0.0    0.0         21.0         0.2    0.0    1  1412.104316   
19  20.0    0.0    0.0         20.0         0.3    0.0    1  1412.104316   
20  21.0    0.0    0.0         19.0         0.3    0.0    1  1412.104316   
21  22.0    0.0    0.0         18.0         0.3    0.0    1  1412.104316   
22  23.0    0.0    0.0         17.0         0.3    0.0    1  1412.104316   
23  24.0    0.0    0.0         16.0         0.3    0.0    1  1412.104316   
24  25.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
25  26.0    0.0    0.0         15.0         0.3    0.0    2  1412.084267   
26  27.0    0.0    0.0         14.0         0.3    0.0    2  1412.084267   
27  28.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
28  29.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
29  30.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
30  31.0    0.0    0.0         12.0         0.3    0.0    2  1412.084267   
31  32.0    0.0    0.0         13.0         0.3    0.0    2  1412.084267   
32  33.0  157.0   79.0         16.0         0.3   78.0    2  1412.084267   
33  34.0  333.0  129.0         20.0         0.2  204.0    2  1412.084267   
34  35.0  509.0  145.0         25.0         0.2  364.0    2  1412.084267   
35  36.0  611.0  169.0         29.0         0.1  442.0    2  1412.084267   
36  37.0  682.0  163.0         30.0         0.1  519.0    2  1412.084267   
37  38.0  667.0  159.0         31.0         0.0  508.0    2  1412.084267   
38  39.0  576.0  149.0         30.0         0.0  427.0    2  1412.084267   
39  40.0  425.0  130.0         28.0         0.0  295.0    2  1412.084267   
40  41.0  234.0   96.0         25.0         0.1  138.0    2  1412.084267   
41  42.0   29.0   28.0         22.0         0.2    1.0    2  1412.084267   
42  43.0    0.0    0.0         21.0         0.2    0.0    2  1412.084267   
43  44.0    0.0    0.0         20.0         0.3    0.0    2  1412.084267   
44  45.0    0.0    0.0         19.0         0.3    0.0    2  1412.084267   
45  46.0    0.0    0.0         18.0         0.3    0.0    2  1412.084267   
46  47.0    0.0    0.0         17.0         0.3    0.0    2  1412.084267   
47  48.0    0.0    0.0         16.0         0.3    0.0    2  1412.084267   
48  49.0    0.0    0.0         16.0         0.3    0.0    3  1412.050859   
49  50.0    0.0    0.0         15.0         0.2    0.0    3  1412.050859   

            δ  daywise_hours  Error in min  LAT_minutes    LAT           ω  \
0  -23.011637              1     -2.904169    19.

In [77]:
Energy_hourly = df['EnergyfromSolar(inJ)'].to_numpy()

In [78]:
print("The amount of energy from solar field is",np.sum(Energy_hourly)/10**12,"in TJ/yr")

The amount of energy from solar field is 3.72928511519894 in TJ/yr
